### ISSUE TITLE
no ingress, nodePort network access between two deployments running on the same provider

### ISSUE THREAD

**Issue by andy108369** [akash-network/support#1]
It appears that Akash deployment network policy blocks the ingress & nodePort access between two deployments running on the same provider.

They are accessible when deployments running on different providers.

ingress & nodePorts (`global: true`) are expected to be open even when deployments are running on the same provider.

**Comment by andy108369** [comment]
Hairpinning might be a solution here as @chainzero mentioned.

**Comment by andy108369** [comment]
This network policy is likely what breaks the communication:

https://github.com/akash-network/provider/blob/v0.2.1/cluster/kube/builder/netpol.go#L123-L138

```
$ kubectl -n $ns get netpol -o yaml
...
...
    - to:
      - ipBlock:
          cidr: 0.0.0.0/0
          except:
          - 10.0.0.0/8
          - 192.168.0.0/16
          - 172.16.0.0/12
```

`10.0.0.0/8` restriction overlaps `kube_service_addresses: 10.233.0.0/18`, `kube_pods_subnet: 10.233.64.0/18`, `calico_pool_cidr: 10.233.64.0/20` - restricting the communication between different deployments.

```
$ ipcalc 10.0.0.0/8
Address:   10.0.0.0             00001010. 00000000.00000000.00000000
Netmask:   255.0.0.0 = 8        11111111. 00000000.00000000.00000000
Wildcard:  0.255.255.255        00000000. 11111111.11111111.11111111
=>
Network:   10.0.0.0/8           00001010. 00000000.00000000.00000000
HostMin:   10.0.0.1             00001010. 00000000.00000000.00000001
HostMax:   10.255.255.254       00001010. 11111111.11111111.11111110
Broadcast: 10.255.255.255       00001010. 11111111.11111111.11111111
Hosts/Net: 16777214              Class A, Private Internet
```

```
kubespray$ git grep -E '^kube_pods_subnet:|^kube_service_addresses:|^calico_pool_cidr:' | column -t | sort -d -k2,2
roles/kubespray-defaults/defaults/main.yaml:kube_service_addresses:              10.233.0.0/18
inventory/sample/group_vars/k8s_cluster/k8s-cluster.yml:kube_service_addresses:  10.233.0.0/18
roles/kubespray-defaults/defaults/main.yaml:kube_pods_subnet:                    10.233.64.0/18
inventory/sample/group_vars/k8s_cluster/k8s-cluster.yml:kube_pods_subnet:        10.233.64.0/18
docs/calico.md:calico_pool_cidr:                                                 10.233.64.0/20
```

Now that this netpol makes it secure when it comes to the pods deployed within different namespace (you don't want someone poking your internal app's ports from his deployment), the only feasible solution I see would be allowing the ingress communication by permitting the ingress-nginx:

```
    - to:
      - namespaceSelector:
          matchLabels:
            kubernetes.io/metadata.name: ingress-nginx
        podSelector: {}
```

**Comment by andy108369** [comment]
I was unable to reproduce either the Ingress or nodePort issues when testing two separate SDLs running on different provider nodes. The tests included H100 and A100 Oblivus, as well as RTX4090 Evergreen providers.

I suspect that this issue may be related to the networking configuration of the provider,, such as an asymmetric routing problem on Valdi provider as described earlier.

For now, I will close this issue until it can be clearly reproduced under standard networking conditions without any special provider configurations that might contribute to the problem.

Regarding the `nodePort` issue previously encountered, I'm left wondering how changing the `akash-deployment-restrictions` resolved the problem at that time. It might have been related to a policy refresh triggered by updating the `akash-deployment-restrictions` network policy.

Alternatively, this issue could also be linked to the following Calico bug, which was fixed in version [`3.27.0`](https://github.com/projectcalico/vpp-dataplane/blob/master/RELEASE_NOTES.md#calico-vpp-v3270). The bug specifically addressed the nodePort service source NAT issue:

```
Fix nodeport service src NAT issue, we do need to src NAT nodeports when the service is not local
```

This fix is not included in Kubespray [v2.24.2](https://github.com/kubernetes-sigs/kubespray/blob/v2.24.2/roles/kubespray-defaults/defaults/main/download.yml#L104) but only gets introduced in Kubespray [v2.25.0](https://github.com/kubernetes-sigs/kubespray/blob/v2.25.0/roles/kubespray-defaults/defaults/main/download.yml#L104).
